packages/jest-haste-map/src/crawlers/watchman.js

```diff
@@ -0,0 +1,112 @@
+ /**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+'use strict';
+
+const H = require('../constants');
+
+const denodeify = require('denodeify');
+const path = require('../fastpath');
+const watchman = require('fb-watchman');
+
+const watchmanURL = 'https://facebook.github.io/watchman/docs/troubleshooting.html';
+
+function isDescendant(root, child) {
+  return child.startsWith(root);
+}
+
+function WatchmanError(error) {
+  return new Error(
+    `Watchman error: ${error.message.trim()}. Make sure watchman ` +
+    `is running for this project. See ${watchmanURL}.`
+  );
+}
+
+module.exports = function watchmanCrawl(
+  roots,
+  extensions,
+  ignore,
+  data
+) {
+  const client = new watchman.Client();
+  const cmd = denodeify(client.command.bind(client));
+  const clocks = data.clocks;
+  let files = data.files;
+  let reset = false;
+
+  return Promise.all(roots.map(root => cmd(['watch-project', root])))
+    .then(responses => {
+      const watchmanRoots = Array.from(
+        new Set(responses.map(response => response.watch))
+      );
+      return Promise.all(watchmanRoots.map(root => {
+        // Build up an expression to filter the output by the relevant roots.
+        const dirExpr = ['anyof'];
+        roots.forEach(subRoot => {
+          if (isDescendant(root, subRoot)) {
+            dirExpr.push(['dirname', path.relative(root, subRoot)]);
+          }
+        });
+        const query = {
+          expression: [
+            'allof',
+            ['type', 'f'],
+            ['anyof'].concat(extensions.map(
+              extension => ['suffix', extension]
+            )),
+            dirExpr,
+          ],
+          fields: ['name', 'exists', 'mtime_ms'],
+        };
+        if (clocks[root]) {
+          // Use the `since` generator if we have a clock available
+          query.since = clocks[root];
+        } else {
+          // Otherwise use the `suffix` generator
+          query.suffix = extensions;
+        }
+
+        return cmd(['query', root, query]).then(response => {
+          if ('warning' in response) {
+            console.warn('watchman warning: ', response.warning);
+          }
+
+          // Reset the file map if watchman was restarted and sends us a list of
+          // files.
+          if (!reset && response.is_fresh_instance) {
+            reset = true;
+            files = Object.create(null);
```

Comment:
yes, that is actually true. The only way to avoid this would be to wait for all queries to finish, then check if one of them is fresh and reset the files object. This might be slower as we'd wait for the slowest operation to finish until we can continue. However, multiple roots doesn't equal multiple watchman roots. In all cases for Facebook (and Jest), all the passed in roots reduce to a single watchman root. Therefore waiting for all queries is not a problem, because it is always just one. I'll make this change.

cc @wez: can it ever happen that watchman gives `is_fresh_instance` for one root but not another? Or will watchman recrawl every watched project if it misses notifications from the file system? I'm assuming when watchman restarts, `is_fresh_instance` would be true for all projects anyway?